### PR TITLE
ENH: rtol and atol for differential_evolution termination fixes #6382

### LINKED ISSUE #6382: optimize.differential_evolution stopping criterion has unintuitive dependence on objective
I've been using optimize.differential_evolution with considerable success but I've noticed a quirk which can greatly perturb the number of iterations and error.  As an example, the following code finds the minimum of the Rosenbrook function (optimize.rosen) offset by a constant using optimize.differential_evolution.  It evaluates each optimization problem multiple times, then prints a table of the offset value, the average number of iterations, and the average error (the squared error on the true minima, at [1, 1]).
```
from __future__ import division

import numpy as np

from scipy.optimize import differential_evolution, rosen

N_VARS = 2
VAR_BOUND = (-5, 5)
REPLICATES = 100

bounds = [VAR_BOUND] * N_VARS

print 'offset\t<iters>\t<error>'

for offset in (0, 1, 10, 100, -10):
	total_iters = 0
	total_error = 0

	for i in xrange(REPLICATES):
		result = differential_evolution(
			lambda x: rosen(x) + offset,
			bounds
			)

		total_iters += result.nit
		total_error += np.sum(np.square(result.x - 1))

	print '{:5n}\t{:6.1f}\t{:0.0e}'.format(
		offset,
		total_iters / REPLICATES,
		total_error / REPLICATES
		)
```
Output:
```
offset	<iters>	<error>
    0	  75.0	1e-19
    1	  20.2	7e-08
   10	  16.5	3e-06
  100	  12.6	2e-05
  -10	  16.8	8e-07
```
The trend should be obvious - the larger the offset, the fewer the number of iterations required to solve.  Likewise, the final error increases by orders of magnitude, though it is still relatively small.  This issue can be tracked to the stopping criterion, which terminates when
```
EPS = np.finfo(np.float64).eps
std(pop)/(mean(pop)+EPS) < tol
```
where "pop" is the vector of the population's objective scores, and "tol" is a parameter supplied to differential_evolution (default is 0.1).  Since the Rosenbrook function has a minimal value of zero, the stopping criterion blows up as the population approaches the minima, only slowing down once the "EPS" term dominates the denominator.

This stopping criterion might make sense if 1) you know for a fact that the minima is zero and 2) you need extremely high precision on your result.  This might make sense for test problems, like the Rosenbrook function.  In general, it's unintuitive and undocumented.

I have two suggestions.  First, replace the current stopping criterion with a simpler function that is insensitive to the population mean.  There is a whole host of possible criterion (see http://link.springer.com/chapter/10.1007%2F978-3-540-68830-3_4#page-1) - I think the standard deviation alone is an excellent default.  Second, allow the user to optionally pass any stopping criterion function they want.

Edit: deleted extraneous code

### PR DESCRIPTION
**Pull request by andyfaff**
Please see #6382 for a history. This PR permits a relative and absolute termination for termination:
1. If `tol` is a single value then terminate when `np.std(population_energies) / np.abs(np.mean(population_energies)) < tol`.
2. If `tol` is a tuple `(rtol, atol)` then terminate when `np.std(population_energies) <= (atol + rtol * np.abs(np.mean(population_energies)))`.

At the moment only a relative tolerance (1) is used. However, if the best solution has `population_energy = 0`, then the denominator on the LHS of (1) tends to zero as the mean approaches zero. This causes the solver to go on for longer than it should. This PR allows an absolute tolerance to be specified as well by passing a tuple `tol=(rtol, atol)`. The default is for `atol=0` to allow back compatibility.

### CODE REVIEW

**Comment by dlax** [comment]
Any reason not to introduce new parameters `rtol` and `atol`? This would be cleaner than handling different types for the `tol` parameter IMHO.

**Comment by pv** [comment]
The optimizers use `tol` for relative tolerance and `atol` for absolute, so maybe just adding `atol` would be enough.

**Comment by andyfaff** [comment]
Agreed. Have reverted `tol` to relative tolerance and added `atol` kwd for absolute tolerance.

File: scipy/optimize/_differentialevolution.py [hunk 1]
```diff
@@ -64,10 +64,10 @@ def differential_evolution(func, bounds, args=(), strategy='best1bin',
         A multiplier for setting the total population size.  The population has
         ``popsize * len(x)`` individuals.
     tol : float, optional
-        When the mean of the population energies, multiplied by tol,
-        divided by the standard deviation of the population energies
-        is greater than 1 the solving process terminates:
-        ``convergence = mean(pop) * tol / stdev(pop) > 1``
+         Relative tolerance for convergence, the solving stops when
```
Comment (dlax) [inline scipy/optimize/_differentialevolution.py]: There appears to be a leading space before `R` which might be the cause refguide build failure.

**Comment by dlax** [comment]
Also could you squash this into a single commit?

**Comment by andyfaff** [comment]
Squashed, let's see if this works now.

**Pull request merged**